sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java

```diff
@@ -234,67 +234,75 @@ private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilde
-                        } else if (typeName.equals("short")) {
-                            propertyHashCalculation = thisProperty.asCondition(" & ", ExpressionDef.constant(0xffff));
-                        } else { // for int and byte, return itself as an int
-                            propertyHashCalculation = thisProperty;
+                            propertyHashCalculation = getPropertyHashValue(propertyTypeDef, propertyGetter);
+                            hashUpdates.add(hashValue.assign(
+                                hashValue.asCondition(" * ", ExpressionDef.constant(HASH_MULTIPLIER)
+                                    .asCondition(" + ", propertyHashCalculation.cast(TypeDef.of(int.class))))));
                         }
-                    } else { // OBJECT
-                        propertyHashCalculation = thisProperty.isNull().asConditionIfElse(
-                            ExpressionDef.constant(NULL_HASH_VALUE),
-                            thisProperty.invoke("hashCode", TypeDef.of(int.class), List.of())
-                        );
-                    }
-                    hashUpdates.add(hashValue.assign(
-                        hashValue.asCondition(" * ", ExpressionDef.constant(HASH_MULTIPLIER)
-                            .asCondition(" + ", propertyHashCalculation.cast(TypeDef.of(int.class))))));
-                }
-                return StatementDef.multi(
-                    parameterDef.get(0).asExpression().isNull().asConditionIf(ExpressionDef.constant(0).returning()),
-                    ((VariableDef.Local) hashValue).defineAndAssign(ExpressionDef.constant(1)),
-                    StatementDef.multi(hashUpdates),
-                    hashValue.returning()
-                );
-            });
+                        hashUpdates.add(hashValue.returning());
+                        return StatementDef.multi(hashUpdates);
+                    })
+                )
+            );
         classDefBuilder.addMethod(method);
     }
+
+    /*
+
+     */
+    private static ExpressionDef getPropertyHashValue(TypeDef propertyTypeDef, ExpressionDef thisProperty) {
+        ExpressionDef propertyHashCalculation;
+        // calculate new property hash value
+        if (propertyTypeDef.isArray()) {
+            String methodName = (((TypeDef.Array) propertyTypeDef).dimensions() > 1) ?  "deepHashCode" : "hashCode";
+            propertyHashCalculation = ClassTypeDef.of(Arrays.class).invokeStatic(methodName, TypeDef.of(int.class), thisProperty);
+        } else if (propertyTypeDef.isPrimitive()) {
+            String typeName = ((TypeDef.Primitive) propertyTypeDef).name();
+            if (propertyTypeDef == TypeDef.BOOLEAN) {
+                propertyHashCalculation = thisProperty.asConditionIfElse(
+                    ExpressionDef.constant(TRUE_HASH_VALUE),
+                    ExpressionDef.constant(FALSE_HASH_VALUE)
+                );
+            } else if (typeName.equals("float")) {
```

Comment:
Please add this primitives as constants to `TypeDef.Primitive` and move the TypeDef.BOOLEAN there, so we have `TypeDef.Primitive.BOOLEAN`